File: Data Structures & Algorithms/course-schedule-ii/submission-6.ts
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {number[]}
     */
    findOrder(numCourses: number, prerequisites: number[][]): number[] {
        const preMap: Record<number, number[]> = {};
        for (const [crs, pre] of prerequisites) {
            if (!(crs in preMap)) preMap[crs] = [];
            preMap[crs].push(pre);
        }
        const cycle: Set<number> = new Set(), visit: Set<number> = new Set(), output: number[] = [];
        const dfs = (crs: number): boolean => {
            if (cycle.has(crs)) return false; // cycle detected
            if (visit.has(crs)) return true; // already in output
            cycle.add(crs);
            if (crs in preMap) {
                for (const pre of preMap[crs]) {
                    if (!dfs(pre)) return false;
                }
            }
            visit.add(crs);
            cycle.delete(crs);
            output.push(crs);
            return true;
        }

        for (let crs = 0; crs < numCourses; crs++) {
            if (!dfs(crs)) return [];
        }
        return output;
    }
}

File: Data Structures & Algorithms/decode-ways/submission-6.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == '0': return 0
        cache = {}
        def dfs(i):
            if i == len(s): return 1
            if i > len(s) or s[i] == '0': return 0
            if i in cache: return cache[i]
            res = dfs(i + 1)
            if i + 1 < len(s) and 10 <= int(s[i:i+2]) <=26: res += dfs(i + 2)
            cache[i] = res
            return cache[i]
        return dfs(0)
